src/Razor/src/Microsoft.AspNetCore.Razor.Language/Syntax/SyntaxNodeExtensions.cs

```diff
@@ -220,5 +239,29 @@ public static TRoot InsertNodesAfter<TRoot>(this TRoot root, SyntaxNode nodeInLi
             var content = string.Concat(tokens.Select(t => t.Content));
             return content;
         }
+
+        private class DiagnosticSyntaxWalker : SyntaxWalker
+        {
+            private readonly List<RazorDiagnostic> _diagnostics;
+
+            public DiagnosticSyntaxWalker()
+            {
+                _diagnostics = new List<RazorDiagnostic>();
+            }
+
+            public IReadOnlyList<RazorDiagnostic> Diagnostics => _diagnostics;
+
+            public override void Visit(SyntaxNode node)
+            {
+                if (node?.ContainsDiagnostics == true)
```

Comment:
@ajaybhargavb from testing, this looks to indicate that the current node or a nested SyntaxNode has diagnostics. Please correct me if I'm wrong here 😄.